crypto/bio/bss_conn.c

```diff
@@ -202,15 +198,27 @@ static int conn_state(BIO *b, BIO_CONNECT *c)
                 ERR_raise_data(ERR_LIB_SYS, i,
                                "calling connect(%s, %s)",
                                 c->param_hostname, c->param_service);
-                BIOerr(BIO_F_CONN_STATE, BIO_R_NBIO_CONNECT_ERROR);
+                if (i == ETIMEDOUT)
+                    BIOerr(BIO_F_CONN_STATE, BIO_R_NBIO_CONNECT_TIMEOUT);
+                else
+                    BIOerr(BIO_F_CONN_STATE, BIO_R_NBIO_CONNECT_ERROR);
                 ret = 0;
                 goto exit_loop;
             } else
                 c->state = BIO_CONN_S_OK;
             break;
 
         case BIO_CONN_S_CONNECT_ERROR:
-            BIOerr(BIO_F_CONN_STATE, BIO_R_CONNECT_ERROR);
```

Comment:
please drop this change.

Reply:
I dropped the differentiation w.r.t. `ETIMEDOUT`, but the move of
```
                   ERR_raise_data(ERR_LIB_SYS, get_last_socket_error(),
                                  "calling connect(%s, %s)",
                                   c->param_hostname, c->param_service);
```
to the new code location is for consistency with the non-blocking case,
and the addition of the guard `i = get_last_socket_error(); if (i != 0)` not only is for consistency but likely also a necessary safeguard. At least for some (good?) reason this guard is present for the non-blocking case.

Reply:
maybe I missed something, but you should never move the errno away from the
place where the error happened, and especially not call a callback in between.

Reply:
All right, I've reverted this.
Likely the guard `i = get_last_socket_error(); if (i != 0)` is not critical.